### PR TITLE
PactDB Checkpointer without version field

### PR DESCRIPTION
**Pull request by emmanueldenloye**
The database schema has been updated to not include the version field any longer.

### CODE REVIEW

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 1]
```diff
@@ -118,9 +117,10 @@ writeSys
     -> v
     -> BlockHandler SQLiteEnv ()
 writeSys wt d k v = do
-    BlockVersion bh version <- gets _bsBlockVersion
+    -- BlockVersion bh _version <- gets _bsBlockVersion
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: Clean up commented-out code

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 2]
```diff
@@ -131,149 +131,143 @@ writeSys wt d k v = do
...
              , SInt (fromIntegral txid)
              ]
+  exec' db mutq [SText tn, SInt (fromIntegral bh)]
   where
+    mutq = "INSERT OR IGNORE INTO VersionedTableMutation VALUES (?,?);"
     q = mconcat [ "UPDATE ["
                 , tn
                 , "] SET rowdata = ? WHERE rowkey = ? AND blockheight = ? \
-                  \ AND version =  ? AND txid = ? ;"
+                  \ AND txid = ? ;"
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: Wait a minute -- shouldn't update produce a new row rather than being a SET operation? If we mutate data tied to an old txid then we are going to break rewind. This is a bug on the original branch too :(

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 3]
```diff
@@ -433,192 +395,163 @@ createBlockHistoryTable =
...
-    vCurrent <- getVCurrent
-    checkHistoryInvariant
-
-    case compare bRestore (bCurrent + 1) of
-        GT -> internalError "handleVersion: Block_Restore invariant violation!"
-        EQ -> newChildBlock bCurrent vCurrent
-        LT -> rewindBlock vCurrent
-
+createTableCreationTable :: BlockHandler SQLiteEnv ()
+createTableCreationTable =
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: local style is 4-space indent, pls fix everywhere

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 4]
```diff
@@ -433,192 +395,163 @@ createBlockHistoryTable =
...
+  callDb "createTableMutationTable" $ \db -> do
+  exec_ db "CREATE TABLE VersionedTableMutation\
+           \(tablename TEXT NOT NULL\
+           \, blockheight UNSIGNED BIGINT NOT NULL\
+           \, CONSTRAINT mutation_unique UNIQUE(tablename,blockheight));"
+  exec_ db "CREATE INDEX mutation_bh ON VersionedTableMutation(blockheight);"
+
+createUserTable :: TableName -> BlockHeight -> BlockHandler SQLiteEnv ()
+createUserTable name bh = do
+  let tablename :: Utf8
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: usually prefer `where` to `let`-blocks in `do` bindings if the values are pure or only depend on the arguments

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 5]
```diff
@@ -433,192 +395,163 @@ createBlockHistoryTable =
...
+      indexname = Utf8 (toS $ asString name <> "_bh")
+  callDb "createUserTable" $ \db -> do
+    exec_ db $ "CREATE TABLE ["
+      <> tablename
+      <> "] (rowkey TEXT\
+           \, blockheight UNSIGNED BIGINT NOT NULL\
+           \, txid UNSIGNED BIGINT NOT NULL\
+           \, rowdata BLOB NOT NULL\
+           \, UNIQUE (blockheight, rowkey, txid));"
+    exec_ db $ "CREATE INDEX [" <> indexname  <> "] ON [" <> tablename <> "](blockheight);"
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: This index is redundant (the `(blockheight, rowkey, txid)` `UNIQUE` constraint is already an index on blockheight), you want a second index on `(rowkey, txid)` instead

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 6]
```diff
@@ -433,192 +395,163 @@ createBlockHistoryTable =
...
+    _ -> internalError "rewindBlock: dropTablesAtRewind: Couldn't resolve the name of the table to drop."
+
+vacuumTablesAtRewind :: BlockHeight -> [Utf8] -> Database -> IO ()
+vacuumTablesAtRewind bh droppedtbls db = do
+  toVacuumTblNames <- qry db "SELECT DISTINCT tablename FROM VersionedTableMutation WHERE blockheight >= ?;"
+    [SInt (fromIntegral bh)]
+    [RText]
+  forM_ toVacuumTblNames $ \case
+    [SText tblname] -> do
+      when (notElem tblname droppedtbls) $
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: `notElem` on list is `O(n)` and this will be called a lot. You have two options for what to return from `dropTablesAtRewind` -- a sorted `Vector` (which you can `bsearch` on using [vector-algorithms](http://hackage.haskell.org/package/vector-algorithms-0.8.0.1/docs/Data-Vector-Algorithms-Search.html)), or a `HashSet Utf8`. `HashSet` gives you the benefit of allowing you to just compute a set difference here and then you don't need the `when` condition.

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 7]
```diff
@@ -433,192 +395,163 @@ createBlockHistoryTable =
...
 deleteHistory = do
-    (BlockVersion bh v) <- gets _bsBlockVersion
+    bh <- gets _bsBlockHeight
     callDb "Deleting from BlockHistory, VersionHistory. Also get txid." $ \db -> do
         exec' db "DELETE FROM BlockHistory WHERE blockheight >= ?"
               [SInt (fromIntegral bh)]
-        exec' db "DELETE FROM VersionHistory WHERE version >= ?;"
-              [SInt (fromIntegral v)]
-        qry db "SELECT endingtxid FROM BlockHistory WHERE blockheight = ?"
+        -- exec' db "DELETE FROM VersionHistory WHERE version >= ?;" [SInt (fromIntegral v)]
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: remove cruft

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 8]
```diff
@@ -433,192 +395,163 @@ createBlockHistoryTable =
...
-               \, rowdata BLOB\
-               \, UNIQUE(blockheight, rowkey, txid));"
+              "CREATE TABLE IF NOT EXISTS ["
+              <> name
+              <> "] (rowkey TEXT\
+                 \, blockheight UNSIGNED BIGINT NOT NULL\
+                 \, txid UNSIGNED BIGINT NOT NULL\
+                 \, rowdata BLOB NOT NULL\
+                 \, UNIQUE(blockheight, rowkey, txid));"
+            exec_ db $ "CREATE INDEX [" <>  indexname <> "] ON [" <> tablename <> "](blockheight);"
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: Same here re blockheight index. What you want instead here is an index on `(rowkey, txid)`. Also, isn't this the same table as above in `createUserTable`? Refactor re "don't-repeat-yourself"

File: src/Chainweb/Pact/Backend/Types.hs [hunk 9]
```diff
@@ -45,7 +45,8 @@ module Chainweb.Pact.Backend.Types
     , bvVersion
     , bvBlock
     , BlockState(..)
-    , bsBlockVersion
+    -- , bsBlockVersion
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/Types.hs]: Fix

File: src/Chainweb/Pact/PactService.hs [hunk 10]
```diff
@@ -163,7 +163,8 @@ initPactService' ver cid chainwebLogger spv bhDb pdb dbDir nodeid resetDb act =
     let loggers = pactLoggers chainwebLogger
     let logger = P.newLogger loggers $ P.LogName ("PactService" <> show cid)
     let gasEnv = P.GasEnv 0 0.0 (P.constGasModel 1)
-    let blockstate = BlockState 0 Nothing (BlockVersion 0 0) M.empty
+    -- let blockstate = BlockState 0 Nothing (BlockVersion 0 0) M.empty
```
Comment (gregorycollins) [inline src/Chainweb/Pact/PactService.hs]: Fix

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 2 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: This was resolved on the branch `feat/pact-replay`. I'll make sure to merge in the latest changes from that branch.

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 3 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 1 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 4 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 5 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 6 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 8 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: Thanks for this suggestion. The code cleans up!

File: src/Chainweb/Pact/Backend/Types.hs (hunk 9 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/Types.hs]: done

File: src/Chainweb/Pact/PactService.hs (hunk 10 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/PactService.hs]: done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 11]
```diff
@@ -355,68 +356,44 @@ doCommit = use bsMode >>= \mm -> case mm of
...
+      err -> internalError $
+        "doGetTxLog: Expected single row with two columns as the \
+        \result, got: " <> T.pack (show err)
   where
-    q = mconcat [ "SELECT rowkey, rowdata FROM ["
+    stmt = mconcat [ "SELECT rowkey, rowdata FROM ["
                 , domainTableName d
-                , "] WHERE txid = ? AND version = ? AND blockheight = ?"
+                , "] WHERE txid = ? AND blockheight = ?"
+
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: spurious newline

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 3 above)
Comment (gregorycollins) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: You say "done" but I see a bunch of 2-space indents below :)

E.g. four spaces after do

```haskell
foo :: Bar -> IO Baz
foo bar = do
    blah
    bar $ do
        etc etc etc
```

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 12]
```diff
@@ -427,194 +404,165 @@ createBlockHistoryTable =
...
+  where
+    createtablestmt =
+      "CREATE TABLE IF NOT EXISTS["
+        <> tablename
+        <> "] (rowkey TEXT\
+             \, blockheight UNSIGNED BIGINT NOT NULL\
+             \, txid UNSIGNED BIGINT NOT NULL\
+             \, rowdata BLOB NOT NULL\
+             \, UNIQUE (blockheight, rowkey, txid));"
+    indexcreationstmt = "CREATE INDEX IF NOT EXISTS [" <> indexname  <> "] ON [" <> tablename <> "](rowkey, txid);"
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: long line. consider `mconcat` instead of `a <> b <> c <> d`

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 13]
```diff
@@ -427,194 +404,165 @@ createBlockHistoryTable =
...
     bCurrent <- getBCurrent
-    vCurrent <- getVCurrent
     checkHistoryInvariant
-
     case compare bRestore (bCurrent + 1) of
-        GT -> internalError "handleVersion: Block_Restore invariant violation!"
-        EQ -> newChildBlock bCurrent vCurrent
-        LT -> rewindBlock vCurrent
-
+      GT -> internalError "handlePossibleRewind: Block_Restore invariant violation!"
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: e.g. you broke the indent here (it was correct before)

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 14]
```diff
@@ -427,194 +404,165 @@ createBlockHistoryTable =
...
-        return $! t
-
-tableMaintenanceRowsVersionedTables :: BlockHandler SQLiteEnv ()
-tableMaintenanceRowsVersionedTables = do
-    tableMaintenanceRowsVersionedSystemTables
-    tableMaintenanceRowsVersionedUserTables
+      assign bsTxId (fromIntegral txid)
+      return $ fromIntegral txid
+      where msg = "handlePossibleRewind: newChildBlock: error finding txid"
+
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: kill blank line

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 15]
```diff
@@ -427,194 +404,165 @@ createBlockHistoryTable =
...
+      _ -> internalError "rewindBlock: dropTablesAtRewind: Couldn't resolve the name of the table to drop."
+  where findTablesToDropStmt =
+          "SELECT tablename FROM VersionedTableCreation where createBlockheight >= ?;"
+
+vacuumTablesAtRewind :: BlockHeight -> HashSet BS.ByteString -> Database -> IO ()
+vacuumTablesAtRewind bh droppedtbls db = do
+    let processMutatedTables ms = fmap (HS.fromList) . forM ms $ \case
+          [SText (Utf8 tbl)] -> return tbl
+          _ -> internalError "rewindBlock: Couldn't resolve the name of the table to possibly vacuum."
+    mutatedTables <- qry db "SELECT DISTINCT tablename FROM VersionedTableMutation WHERE blockheight >= ?;"
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: long line

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 11 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: addressed

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 12 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done & done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 13 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done & done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 14 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done & done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 15 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done & done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 16]
```diff
@@ -131,99 +130,99 @@ writeSys wt d k v = do
...
-backendWriteInsert key tn bh version txid v db =
+backendWriteInsert key tn bh txid v db = do
     exec' db q [ SText key
                , SInt (fromIntegral bh)
-               , SInt (fromIntegral version)
                , SInt (fromIntegral txid)
                , SBlob (toStrict (Data.Aeson.encode v))]
+    exec' db mutq [SText tn, SInt (fromIntegral bh)]
   where
+    mutq = "INSERT OR IGNORE INTO VersionedTableMutation VALUES (?,?);"
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: note for possible performance followup -- this is a place where a per-block cache would be appropriate, so we don't keep ramming sqlite

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 17]
```diff
@@ -131,99 +130,99 @@ writeSys wt d k v = do
...
-             ]
+backendWriteUpdate key tn bh txid v db = do
+    exec' db q [ SText key
+               , SInt (fromIntegral bh)
+               , SInt (fromIntegral txid)
+               , SBlob (toStrict (Data.Aeson.encode v))
+               ]
+    exec' db mutq [SText tn, SInt (fromIntegral bh)]
   where
+    mutq = "INSERT OR IGNORE INTO VersionedTableMutation VALUES (?,?);"
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: Dupe code (I see at least three), refactor into function

```haskell
markMutation :: Utf8 -> BlockHeight -> Database -> IO ()
```

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 18]
```diff
@@ -427,194 +396,174 @@ createBlockHistoryTable =
...
+        -- enforce invariant that the history has
+        -- (B_restore-1,H_parent).
+        historyInvariant <- callDb "handlePossibleRewind" $ \db -> do
             qry db "SELECT COUNT(*) FROM BlockHistory WHERE \
-                   \blockheight = ? AND hash = ?;"
-                [ SInt $! fromIntegral $ pred bRestore
-                , SBlob (Data.Serialize.encode hsh) ]
-                [RInt]
-            >>= expectSingleRowCol "handleVersion: (historyInvariant):"
+                   \blockheight = ? and hash = ?;"
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: revert this change, sql keywords are uppercase in all our queries

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 19]
```diff
@@ -427,194 +396,174 @@ createBlockHistoryTable =
...
+      exec_ db "CREATE TABLE IF NOT EXISTS VersionedTableMutation\
+               \(tablename TEXT NOT NULL\
+               \, blockheight UNSIGNED BIGINT NOT NULL\
+               \, CONSTRAINT mutation_unique UNIQUE(tablename,blockheight));"
+      exec_ db "CREATE INDEX IF NOT EXISTS mutation_bh ON VersionedTableMutation(blockheight);"
+
+createUserTable :: TableName -> BlockHeight -> BlockHandler SQLiteEnv ()
+createUserTable name bh =
+    callDb "createUserTable" $ \db -> do
+      createVersionedTable tablename indexname db
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: note two-char indent here

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 20]
```diff
@@ -427,194 +396,174 @@ createBlockHistoryTable =
...
+    callDb "createTableCreationTable" $ \db -> exec_ db
+      "CREATE TABLE IF NOT EXISTS VersionedTableCreation\
+      \(tablename TEXT NOT NULL\
+      \, createBlockheight UNSIGNED BIGINT NOT NULL\
+      \, CONSTRAINT creation_unique UNIQUE(tablename, createBlockheight));"
+
+createTableMutationTable :: BlockHandler SQLiteEnv ()
+createTableMutationTable =
+    callDb "createTableMutationTable" $ \db -> do
+      exec_ db "CREATE TABLE IF NOT EXISTS VersionedTableMutation\
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: two char indent here

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 21]
```diff
@@ -427,194 +396,174 @@ createBlockHistoryTable =
...
-    tableMaintenanceRowsVersionedSystemTables
-    tableMaintenanceRowsVersionedUserTables
+dropTablesAtRewind :: BlockHeight -> Database -> IO (HashSet BS.ByteString)
+dropTablesAtRewind bh db = do
+    toDropTblNames <- qry db findTablesToDropStmt [SInt (fromIntegral bh)] [RText]
+    tbls <- fmap (HS.fromList) . forM toDropTblNames $ \case
+        [SText tblname@(Utf8 tbl)] -> do
+            exec_ db $ "DROP TABLE " <> tblname <> ";"
+            return tbl
+        _ -> internalError "rewindBlock: dropTablesAtRewind: Couldn't resolve the name of the table to drop."
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: wrap long lines (>= 79 col) in this function

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 16 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: understood. I'll keep this in mind.

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 17 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done and done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 18 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done and done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 19 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done and done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 20 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done and done

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs (hunk 21 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done and done

Comment (emmanueldenloye) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: done and done
